pkg/portal/transport/stripe_webhook_handler.go

```diff
@@ -0,0 +1,180 @@
+package transport
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/authgear/authgear-server/pkg/lib/infra/db/globaldb"
+	"github.com/authgear/authgear-server/pkg/portal/libstripe"
+	"github.com/authgear/authgear-server/pkg/portal/model"
+	"github.com/authgear/authgear-server/pkg/portal/service"
+	"github.com/authgear/authgear-server/pkg/util/httproute"
+	"github.com/authgear/authgear-server/pkg/util/log"
+)
+
+func ConfigureStripeWebhookRoute(route httproute.Route) httproute.Route {
+	return route.WithMethods("POST").WithPathPattern("/api/subscription/webhook/stripe")
+}
+
+type StripeWebhookLogger struct{ *log.Logger }
+
+func NewStripeWebhookLogger(lf *log.Factory) StripeWebhookLogger {
+	return StripeWebhookLogger{lf.New("stripe-webhook")}
+}
+
+type StripeService interface {
+	ConstructEvent(r *http.Request) (libstripe.Event, error)
+	CreateSubscriptionIfNotExists(stripeCheckoutSessionID string) error
+}
+
+type SubscriptionService interface {
+	UpdateSubscriptionCheckoutStatusAndCustomerID(appID string, stripCheckoutSessionID string, status model.SubscriptionCheckoutStatus, customerID string) error
+	UpdateSubscriptionCheckoutStatusByCustomerID(appID string, customerID string, status model.SubscriptionCheckoutStatus) error
+	CreateSubscription(appID string, stripeSubscriptionID string, stripeCustomerID string) (*model.Subscription, error)
+	UpdateAppPlan(appID string, planName string) error
+}
+
+type StripeWebhookHandler struct {
+	StripeService StripeService
+	Logger        StripeWebhookLogger
+	Subscriptions SubscriptionService
+	Database      *globaldb.Handle
+}
+
+func (h *StripeWebhookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var err error
+	defer func() {
+		if err != nil {
+			h.Logger.WithError(err).Errorf("failed to handle stripe webhook")
+			http.Error(w, "failed to handle stripe webhook", http.StatusBadRequest)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	}()
+
+	event, err := h.StripeService.ConstructEvent(r)
+	if err != nil {
+		if errors.Is(err, libstripe.ErrUnknownEvent) {
+			// It is common to receive unknown event
+			// e.g. create objects via stripe portal doesn't have the metadata
+			//      event type that are not handled by the server
+			// gracefully ignore them
+			err = nil
+			w.WriteHeader(http.StatusOK)
```

Comment:
This will cause header to be written twice because we have a defer block above to write header.

```golang
defer func() {
  if errors.Is(err, libstripe.ErrUnknownEvent) {
    w.WriteHeader(http.StatusOK)
    return
  }
  if err != nil {
    h.Logger.WithError(err).Errorf("failed to handle stripe webhook")
    http.Error(w, "failed to handle stripe webhook", http.StatusBadRequest)
    return
  }
  w.WriteHeader(http.StatusOK)
}()

event, err := h.StripeService.ConstructEvent(r)
if err != nil {
  return
}
```